src/arch/arm/config.cmake

```diff
@@ -40,7 +40,7 @@ config_set(KernelArmICacheVIPT ARM_ICACHE_VIPT "${KernelArmICacheVIPT}")
 if(KernelSel4ArchAarch32)
     # 64-bit targets may be building in 32-bit mode,
     # so make sure maximum paddr is 32-bit.
-    math(EXPR KernelPaddrUserTop "(1 << 32) - 1")
+    set(KernelPhysAddressSpaceBits 32)
```

Comment:
The `1 << 32` is undefined behaviour in C when you're in u32. This is also the reason, the kernel uses the `2^n-1` form in general, i.e. talks about the last valid address and compares with `<=`, not the one after and comparing with `<` (not just here, for other comparisons as well). The former is always safe wrt overflow and you don't have to think (or prove anything) about it, whereas `+1` is an operation that can overflow, and then `<` can have unexpected behaviour.

So if we standardise, we should standardise on `(1 << n) - 1`

Reply:
We can take care of this in the C code then also and there is a compiler warning anyway if the shift exceeds the limits. But having the address space bit size information available give much more flexibility than having some address and then try deriving the phys address space size in bits from that somehow.

Reply:
Doing that in C is awkward. But It would be perfectly fine to configure the bit size as you have it, and then compute `KernelPaddrUserTop` from it. Then you have both available.

Reply:
I have brought `KernelPaddrUserTop` back, but this time in the generic CMake code, so there is only one place where to maintain this